src/Conversion/ONNXToTOSA/Math/Gemm.cpp

```diff
@@ -188,19 +188,25 @@ class ONNXGemmOpLoweringToTOSA : public OpConversionPattern<ONNXGemmOp> {
     // If all check passed, we replace the GEMM by a FC operator
     Type resultType = getTypeConverter()->convertType(op.getResult().getType());
 
-    // If no bias is given to the GEMM operator, we create a 1D bias with all
-    // zeroes
-    if (C.getType().isa<mlir::NoneType>()) {
-      // Base C format on B[0] format
-      ArrayRef<int64_t> cformat(B.getType().cast<TensorType>().getShape()[0]);
-      std::vector<float> elements = {};
-      for (int i = 0; i < cformat[0]; ++i)
-        elements.push_back(0.0F);
-      C = tosa::getConstTensor<float>(rewriter, op, elements, cformat).value();
+    // Because the bias is not broadcastable for TOSA while it is for Onnx,
+    // we create an empty bias and use an add (broadcastable for tosa)
+    // afterwards
+    // Base dummy C shape on B[0] shape
+    ArrayRef<int64_t> cformat(B.getType().cast<TensorType>().getShape()[0]);
+    std::vector<float> elements = {};
+    for (int i = 0; i < cformat[0]; ++i)
+      elements.push_back(0.0F);
+    Value dummyC = tosa::getConstTensor<float>(rewriter, op, elements, cformat).value();
+
+    Value fcRes = tosa::CreateOpAndInfer<mlir::tosa::FullyConnectedOp>(rewriter,
+        op->getLoc(), resultType, A, B, dummyC).getResult();
+    // If C was present in the original GEMM, we create an add to take the bias into account
+    if (!C.getType().isa<mlir::NoneType>()) {
+      tosa::CreateReplaceOpAndInfer<mlir::tosa::AddOp>(rewriter, op, resultType, fcRes, C);
+    }
```

Comment:
My other comment would basically mean to have three cases here:

- The none case
- The correct case
- The broadcast case

Reply:
What do you call the correct case? I'm not sure I get what your first comment means. Because of the requirements of bias having to be of rank 1 to be lowered into an FC, aren't all the case that are not the `none` case in the `broadcast` case?

Reply:
for example in the lit test you changed. There would be 3 ways in onnx to handle it. `none`, `tensor<1xf32>` and `tensor<4xf32>`. The last one can just use a `fully_connected`, the first one needs to create a tensor<4xf32> and then a `fully_connected` and the middle one needs this `tosa.add`. I think it is better to separate the cases, because then you have less created ops.